crates/validator-transaction-pool/src/lib.rs

```diff
@@ -27,104 +22,125 @@ impl TransactionFilter {
-) -> (ReadClient, Vec<SingleTopicWriteClient>) {
-    let topics: Vec<Topic> = topic_tx_pairs.iter().map(|(topic, _)| *topic).collect();
-    let pool_state = Arc::new(Mutex::new(PoolState::new(topic_tx_pairs)));
-    let read_client = ReadClient {
-        pool: pool_state.clone(),
-    };
-    let write_clients = topics
-        .into_iter()
-        .map(|topic| SingleTopicWriteClient {
-            pool: pool_state.clone(),
-            topic,
-        })
-        .collect();
-    (read_client, write_clients)
+pub type VTxnPoolWrapper = Arc<Mutex<PoolState>>;
+
+/// Create a new validator transaction pool.
+pub fn new() -> VTxnPoolWrapper {
+    Arc::new(Mutex::new(PoolState::default()))
+}
+
+struct PoolItem {
+    seq_num: u64,
+    topic: Topic,
+    txn: Arc<ValidatorTransaction>,
+    pull_notification_tx: Option<aptos_channel::Sender<(), Arc<ValidatorTransaction>>>,
 }
 
-struct PoolState {
-    /// sorted by priority (high to low).
-    topics: Vec<Topic>,
+/// PoolState invariants.
+/// `(seq_num=i, topic=T)` exists in `txn_queue` if and only if it exists in `seq_nums_by_topic`.
+#[derive(Default)]
+pub struct PoolState {
+    /// Incremented every time a txn is pushed in. The txn gets the old value as its sequence number.
+    next_seq_num: u64,
 
-    /// Currently only support 1 txn per topic.
-    txns: HashMap<Topic, Arc<ValidatorTransaction>>,
+    /// Track Topic -> seq_num mapping.
+    /// We allow only 1 txn per topic and this index helps find the old txn when adding a new one for the same topic.
+    seq_nums_by_topic: HashMap<Topic, u64>,
 
-    pull_notification_senders: HashMap<Topic, Mutex<PullNotificationSender>>,
+    /// Txns ordered by their sequence numbers (i.e. time they entered the pool).
+    txn_queue: BTreeMap<u64, PoolItem>,
+}
+
+/// Returned for `txn` when you call `PoolState::put(txn, ...)`.
+/// If this is dropped, `txn` will be deleted from the pool (if it has not been).
+///
+/// This allows the pool to be emptied on epoch boundaries.
+pub struct TxnGuard {
+    pool: Arc<Mutex<PoolState>>,
+    seq_num: u64,
 }
 
 impl PoolState {
-    pub fn new(topic_sender_pairs: Vec<(Topic, Option<PullNotificationSender>)>) -> Self {
-        let topics: Vec<Topic> = topic_sender_pairs.iter().map(|(topic, _)| *topic).collect();
-        let topic_set: HashSet<Topic> = HashSet::from_iter(topics.clone());
-        assert_eq!(topics.len(), topic_set.len());
-        let pull_notification_senders = topic_sender_pairs
-            .into_iter()
-            .filter_map(|(topic, maybe_sender)| {
-                maybe_sender.map(|sender| (topic, Mutex::new(sender)))
-            })
-            .collect();
-        Self {
-            txns: HashMap::new(),
-            pull_notification_senders,
-            topics,
+    /// Append a txn to the pool.
+    /// Return a txn guard that allows you to later delete the txn from the pool.
+    pub fn put(
+        pool: Arc<Mutex<Self>>,
```

Comment:
this looks weird, it probably should be implemented for the outer instead of the inner

Reply:
fixed